check_pgactivity

```diff
@@ -3572,6 +3575,85 @@ sub check_database_size {
 }
 
 
+=item B<extensions_versions> (9.1+)
+
+Check all extensions installed in all databases (including templates)
+and raise a critical alert if the current version is not the default
+version available on the instance (according to pg_available_extensions).
+
+Typically, it is used to detect forgotten extension upgrades after package
+upgrades or a pg_upgrade.
+
+Perfdata returns the number of outdated extensions in each database.
+
+This service supports both C<--dbexclude> and C<--dbinclude> parameters.
+
+This service supports multiple C<--exclude> argument to exclude one or
+more extensions from the check.
+
+Required privileges: unprivileged role able to log in all databases
+
+=cut
+sub check_extensions_versions {
+    my @rs;
+    my @perfdata;
+    my @msg;
+    my @longmsg;
+    my @hosts;
+    my @all_db;
+    my $nb;
+    my $me           = 'POSTGRES_CHECK_EXT_VERSIONS';
+    my %args         = %{ $_[0] };
+    my @dbinclude    = @{ $args{'dbinclude'} };
+    my @dbexclude    = @{ $args{'dbexclude'} };
+    my $tot_outdated = 0 ;
+    my $query = q{SELECT name, default_version, installed_version
+                 FROM pg_catalog.pg_available_extensions
+                 WHERE installed_version != default_version};
+
+    @hosts = @{ parse_hosts %args };
+
+    pod2usage(
+        -message => 'FATAL: you must give only one host with service "extensions_versions".',
+        -exitval => 127
+    ) if @hosts != 1;
+
+    is_compat $hosts[0], 'extensions_versions', $PG_VERSION_91 or exit 1;
+
+    @all_db = @{ get_all_dbname( $hosts[0], 'all_dbs' ) };
+
+    # Iterate over all db
+    ALLDB_LOOP: foreach my $db (sort @all_db) {
+        next ALLDB_LOOP if grep { $db =~ /$_/ } @dbexclude;
+        next ALLDB_LOOP if @dbinclude and not grep { $db =~ /$_/ } @dbinclude;
+
+        my $outdated = 0;
+
+        # For each record: extension, default, installed
+        @rs = @{ query ( $hosts[0], $query, $db ) };
+
+        REC_LOOP: foreach my $ext (sort @rs) {
+            foreach my $exclude_re ( @{ $args{'exclude'} } ) {
```

Comment:
I still think that this is too simplistic.  What if for some reason you can't upgrade an extension in only one (or a subset) of your databases but want to check that it's up-do-date in all others?

Reply:
what do you propose as user interaction for such feature?

Reply:
Same as what I suggested in May, ie. a way to specify extname or (extname and dbname).  Maybe something like
 * `--exclude 'extname'`
 * `--exclude 'dbname/extname'`

Reply:
well, why not. I'll do it during the week.

Reply:
This PR is old enough, so I've added what seems to be the last missing part about `--exclude 'dbname/extname'`, and a few comments: 4f255cc25310e27915dceb896aa9a00997612862

Example:
```bash
$ check_pgactivity --service extensions_versions --dbservice=nagios_defo12 

POSTGRES_CHECK_EXT_VERSIONS CRITICAL | bis=2;;1;0 gin=2;;1;0 partitions_100_par_fdw=0;;1;0 postgres=2;;1;0 scratch=2;;1;0 template1=0;;1;0 textes10=0;;1;0 tpc=1;;1;0
bis.unit: 5 (should be: 7) bis.nulle: 0.0.2 (should be: 0.0.3) gin.nulle: 0.0.2 (should be: 0.0.3) gin.unit: 3 (should be: 7) postgres.unit: 3 (should be: 7) postgres.nulle: 0.0.2 (should be: 0.0.3) scratch.unit: 4 (should be: 7) scratch.nulle: 0.0.2 (should be: 0.0.3) tpc.nulle: 0.0.1 (should be: 0.0.3)

$ check_pgactivity --service extensions_versions --dbservice=nagios_defo12  --exclude 'scratch/nulle' --exclude 'bis/nulle' --exclude 'gin/nulle' --exclude 'unit'  --dbexclude postgres

POSTGRES_CHECK_EXT_VERSIONS CRITICAL | bis=0;;1;0 gin=0;;1;0 partitions_100_par_fdw=0;;1;0 scratch=0;;1;0 template1=0;;1;0 textes10=0;;1;0 tpc=1;;1;0
tpc.nulle: 0.0.1 (should be: 0.0.3)

$ check_pgactivity --service extensions_versions --dbservice=nagios_defo12  --exclude 'scratch/nulle' --exclude 'bis/nulle' --exclude 'gin/nulle' --dbexclude postgres --exclude 'tpc/nulle' --exclude 'unit'

POSTGRES_CHECK_EXT_VERSIONS OK | bis=0;;1;0 gin=0;;1;0 partitions_100_par_fdw=0;;1;0 scratch=0;;1;0 template1=0;;1;0 textes10=0;;1;0 tpc=0;;1;0
```

Reply:
Up! @rjuju are you okay?

Reply:
I did a code review and fix a small incompatibility.
The patch is ok to squash & merge imo.
@rjuju: one last review round/red flag before I merge?

Reply:
thanks a lot, it all looks good to me!